channeldb/db.go

```diff
@@ -147,32 +146,35 @@ type DB struct {
 	clock  clock.Clock
 }
 
-// Open opens an existing channeldb. Any necessary schemas migrations due to
-// updates will take place as necessary.
+// Open opens or creates channeldb. Any necessary schemas migrations due
+// to updates will take place as necessary.
 func Open(dbPath string, modifiers ...OptionModifier) (*DB, error) {
-	path := filepath.Join(dbPath, dbName)
+	backend, err := kvdb.GetBoltBackend(dbPath, dbName)
+	if err != nil {
+		return nil, err
+	}
 
-	if !fileExists(path) {
-		if err := createChannelDB(dbPath); err != nil {
-			return nil, err
-		}
+	db, err := OpenWithBackend(backend, modifiers...)
+	if err == nil {
+		db.dbPath = dbPath
+	}
+	return db, err
+}
+
+// OpenWithBackend opens or creates channeldb using the passed kvdb.Backend.
+// Any necessary schemas migrations due to updates will take place as necessary.
+func OpenWithBackend(backend kvdb.Backend, modifiers ...OptionModifier) (*DB, error) {
+	if err := initChannelDB(backend); err != nil {
```

Comment:
As is this, more or less assumes the backend is already "open", as it just does initialization. What about just passing in the args and name, then using `kvdb.Open(name, args...)`? We'd get a`kvdb.Backend` instance out of it, then can proceed with set up as normal.

Reply:
The main reason this is done as this way is to support testing with multiple backends without much code rewrite. So when we test we first instantiate the `kvdb.Backend` then "open" the channeldb.
Another strong reason is that the `kvdb.Backend` becomes a dependency this way and it's much simpler to stub/mock it if needed.